### PR TITLE
refactor tests in expression evaluator and improve support for expression evaluation when we missing classpath (e.g remotedebugging)

### PR DESCRIPTION
**Pull request by romanowski**
Refactor test in expression evaluator:
- create single project containing all ee-related sources
- compile and setup project only once for all test - speedup

EE in Remote without classpath (In all cases we don't have classpath):
- Add support for values from this and other this components.
- Add support for methods and var from this scope (including parents)
- Create remote test suite to test working/non working features.

### CODE REVIEW

**Comment by Kwestor** [comment]
Ok to test

**Comment by dragos** [comment]
ok to test

**Comment by dragos** [comment]
add to whitelist

File: org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/CachedProjectContext.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package org.scalaide.debug.internal
+package expression
+
+import org.junit.AfterClass
+import org.junit.BeforeClass
+
+/**
+ * Caches eclipse project for whole test case.
+ *  Project is started before target class and cleaned at the end.
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/CachedProjectContext.scala]: Reformat (2 spaces before text).

File: org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/EclipseProjectContext.scala [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import org.scalaide.core.testsetup.SDTTestUtils
+import org.scalaide.core.testsetup.TestProjectSetup
+import org.scalaide.debug.internal.ScalaDebugRunningTest
+import org.scalaide.debug.internal.ScalaDebugTestSession
+import org.scalaide.debug.internal.launching.RemoteConnector
+import org.eclipse.debug.core.ILaunchConfiguration
+import collection.mutable
+import collection.JavaConverters._
+
+
```
Comment (pkukielka) [inline org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/EclipseProjectContext.scala]: Remove new lines

File: org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/EclipseProjectContext.scala [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.eclipse.debug.core.DebugPlugin
+import org.junit.Assert.fail
+import org.scalaide.core.testsetup.SDTTestUtils
+import org.scalaide.core.testsetup.TestProjectSetup
+import org.scalaide.debug.internal.ScalaDebugRunningTest
+import org.scalaide.debug.internal.ScalaDebugTestSession
+import org.scalaide.debug.internal.launching.RemoteConnector
+import org.eclipse.debug.core.ILaunchConfiguration
+import collection.mutable
+import collection.JavaConverters._
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/EclipseProjectContext.scala]: Use `scala.collection.`

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/Transformers.scala [hunk 4]
```diff
@@ -96,15 +100,81 @@ abstract class AstTransformer[+Tpe <: TypecheckRelation]
...
-        NewInstanceCall(className, transformSingleTree(argList, super.transform))
+        NewInstanceCall(className, transform(argList))
       case tree =>
-        transformSingleTree(tree, super.transform)
+        transformSingleTree(tree, baseTransform)
+    }
+
+    import Scala._
+
+    /** Match exact select dynamic call.
```
Comment (pkukielka) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/Transformers.scala]: Fix formatting please

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/phases/MockUnboundValuesAndAddImportsFromThis.scala [hunk 5]
```diff
@@ -35,6 +35,19 @@ import org.scalaide.logging.HasLogger
...
+ * Than this code:
+ * {{{
+ *    method(value)
+ * }}}
+ * Becomes:
+ * {{{
+ *  val _this: JdiProxy = JdiContext.placeholder
+ *  import __this._
+ *  val value = _this.value
+ *  def method(v1: Any) = _this.method(v1)
```
Comment (pkukielka) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/phases/MockUnboundValuesAndAddImportsFromThis.scala]: one line is missing: `method(value)`

File: org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/EclipseProjectContext.scala [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    setup.launchInRunMode(fileName, port, testStartDelay)
+    Thread.sleep(remoteConnectionDelay) //to be sure that project is started
+    operation(ScalaDebugTestSession(createRemoteConnection(port)))
+  }
+
+  private def createRemoteConnection(port: Int): ILaunchConfiguration = {
+    val manager = DebugPlugin.getDefault.getLaunchManager
+
+    val remoteCategory = manager.getLaunchConfigurationType("org.eclipse.jdt.launching.remoteJavaApplication")
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/EclipseProjectContext.scala]: Move those 4 strings to `val`s.

File: org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/remote/RemoteTestCase.scala [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2015 Contributor. All rights reserved.
+ */
+package org.scalaide.debug.internal.expression.remote
+
+import org.junit.BeforeClass
+import org.junit.AfterClass
+import org.scalaide.debug.internal.expression.EclipseProjectContext
+
+trait RemoteTestCase {
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/remote/RemoteTestCase.scala]:
```scala
/** Mix this trait to make your test class run in remote context. */
```

File: org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/remote/WithoutProjectSetTest.scala [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  @Test
+  def expectReflectiveCompilationError(): Unit =
+    expectReflectiveCompilationError(""" list.map( _ + int) """)
+
+  @Test
+  def typedLambdaWithClosure(): Unit =
+    eval(""" list.map( (i: Int)=> i + int) """, list.map((i: Int) => i + 1), Scala.::)
+
+  @Ignore("TODO O-8791 - No class def found is thrown")
+  @Test
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression.tests/src/org/scalaide/debug/internal/expression/remote/WithoutProjectSetTest.scala]: Could you change `@Ignore` to `@Test(expected = classOf[NoClassDefFound]`)? That way if error changes in future (or is no longer there) we will get feedback.

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/ExpressionException.scala [hunk 9]
```diff
@@ -59,8 +59,12 @@ object ExpressionException {
 
   def multipleMethodsMatchNestedOneMessage(methodName: String, candidates: Seq[String]) =
     s"""We cannot determine real name of nested method: $methodName}.
-       |Candidates: ${candidates.mkString(", ")}
-       |Nested methods are not fully supported in expression evaluator.")""".stripMargin
+        |Candidates: ${candidates.mkString(", ")}
+        |Nested methods are not fully supported in expression evaluator.")""".stripMargin
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/ExpressionException.scala]: Imo this one additional space is unnecessary.

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/ExpressionManager.scala [hunk 10]
```diff
@@ -111,7 +125,7 @@ trait ExpressionManager extends HasLogger {
               SuccessWithoutValue(outputText)
           }
         case Failure(exception) =>
-          val errorMessage = exception.getMessage
+          val errorMessage = renderErrorMessage(context.hasClasspath, exception)
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/ExpressionManager.scala]: Pass `exception.getMessage` to `renderErrorMessage`.

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/Transformers.scala [hunk 11]
```diff
@@ -96,15 +100,81 @@ abstract class AstTransformer[+Tpe <: TypecheckRelation]
...
+        Apply(Select(tree, TermName(SelectDynamicName)), List(name))
+
+      def unapply(t: Tree): Option[(Tree, Tree)] = t match {
+        case Apply(Select(on, TermName(SelectDynamicName)), List(name)) =>
+          Some((on, name))
+        case _ => None
+      }
     }
+
+    /** Match exact update dynamic call.
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/Transformers.scala]: Reformat doc comments (all 4).

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/context/extensions/ExtendedContext.scala [hunk 12]
```diff
@@ -30,9 +26,35 @@ case class ExtendedContext(currentFrame: StackFrame)
...
   /** Get type for variable - if variable is one of 'this' variables returns its type */
-  final def typeFor(name: universe.TermName): Option[ReferenceType] =
+  final def typeFor(name: universe.TermName): Option[Type] =
     currentTransformation.flatMap(_.elementMap.get(name)).map(_.referenceType)
 
+  /**
+   * Search for member using JDI in each of this element.
+   * This used only when we missing classpath (e.g. Remote debugging).
+   */
+  final def findMember(variable: Variable): Option[TypedVariable] =
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/context/extensions/ExtendedContext.scala]: This **is** used...

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/context/extensions/ExtendedContext.scala [hunk 13]
```diff
@@ -30,9 +26,35 @@ case class ExtendedContext(currentFrame: StackFrame)
       .map(parentClasses)
 
   /** Get type for variable - if variable is one of 'this' variables returns its type */
-  final def typeFor(name: universe.TermName): Option[ReferenceType] =
+  final def typeFor(name: universe.TermName): Option[Type] =
     currentTransformation.flatMap(_.elementMap.get(name)).map(_.referenceType)
 
+  /**
+   * Search for member using JDI in each of this element.
+   * This used only when we missing classpath (e.g. Remote debugging).
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/context/extensions/ExtendedContext.scala]: This **is** used ...

File: org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/phases/MockUnboundValuesAndAddImportsFromThis.scala [hunk 14]
```diff
@@ -113,13 +130,32 @@ class MockUnboundValuesAndAddImportsFromThis(val toolbox: ToolBox[universe.type]
...
+        case PlainTypedVariable(_, typeName) =>
           typeName
-      }.map(valueDefinitionCode(variable))
+        case dynamic => throw new RuntimeException("Cannot build static proxy for dynamic variable: " + dynamic)
+      }
+
+      valueDefinitionCode(variable.untyped)(typeName)
+    }
+
+    /** Build dynamic member variable implementation.
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.expression/src/org/scalaide/debug/internal/expression/proxies/phases/MockUnboundValuesAndAddImportsFromThis.scala]: Reformat doc comment.

File: org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/RemoteConnectorTest.scala [hunk 15]
```diff
@@ -20,14 +18,29 @@ import org.eclipse.debug.core.ILaunchManager
...
 import org.junit.Test
+import org.scalaide.core.testsetup.TestProjectSetup
 import org.scalaide.debug.internal.EclipseDebugEvent
 import org.junit.Ignore
+import org.eclipse.debug.core.IStreamListener
+import org.eclipse.debug.core.model.IStreamMonitor
+
+import org.scalaide.debug.internal.ScalaDebugRunningTest
+import org.scalaide.debug.internal.ScalaDebugTestSession
+import org.scalaide.debug.internal.ScalaDebugTestSession.addDebugEventListener
```
Comment (Kwestor) [inline org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/RemoteConnectorTest.scala]: Organize

**Comment by Kwestor** [comment]
Review done.

**Comment by Kwestor** [comment]
LGTM, @romanowski, did you check this failing test?

**Comment by sschaef** [comment]
Rebase required.

**Comment by Kwestor** [comment]
Closing this for now, it needs some more work.

**Pull request closed without merging**